include/counter.h

```diff
@@ -22,35 +22,106 @@
 #include <zephyr/types.h>
 #include <stddef.h>
 #include <device.h>
+#include <stdbool.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-typedef void (*counter_callback_t)(struct device *dev, void *user_data);
+struct counter_alarm_cfg;
+
+/** @brief Alarm callback
+ *
+ * @param dev     Pointer to the device structure for the driver instance.
+ * @param cb      Original structure owning this handler.
+ * @param counter Counter value that triggered the callback.
+ */
+typedef void (*counter_alarm_callback_t)(struct device *dev,
+					 struct counter_alarm_cfg *cb,
+					 u32_t counter);
+
+/** @brief Alarm callback structure.
+ *
+ * Used to set an alarm. *
+ * Beware such structure should not be allocated on stack.
+ */
+struct counter_alarm_cfg {
+	u8_t channel_id; /*!< Channel ID. Number of channels is driver
+			  * dependent.
+			  */
+
+	bool absolute; /*!< If true ticks are treated as absolute value, else
+			*   it is relative to the counter reading performed
+			*   during the call.
+			*/
+
+	u32_t ticks; /*!< In case of absolute flag is set, maximal value that
+		      *   can be set equals wrap value (counter_get_wrap).
+		      *   Otherwise counter_get_max_relative_alarm() returns
+		      *   maximal value that can be set. If counter is clock
+		      *   driven then ticks can be converted to microseconds
+		      *   (see @ref counter_ticks_to_us). Alternatively,
+		      *   counter implementation may count asynchronous events.
+		      */
+
+	counter_alarm_callback_t handler; /*!< handler called on alarm
+					   *   (cannot be NULL).
+					   */
+};
+
+typedef void (*counter_wrap_callback_t)(struct device *dev, void *user_data);
```

Comment:
I am not convinced the `user_data` is necessary here. Do you see a use case for this?
If one would like to use a common wrap callback for several devices, any "user data" can be placed in the device data structure. And for a particular device the wrap callback function cannot be used for anything else (the alarm callback is incompatible), so I don't see a need to have an additional parameter that somehow distinguishes calls.

Reply:
what if there is a multi-instance module which is using counter api and handles wrap callback. Isn't user_data the simplest way of getting back handle to the object that used the counter?

Reply:
Now I see I mixed up a few things. Obviously, the application cannot use anything pointed from the `dev` structure to get the object associated with the callback call. Sorry, my bad.
But still, my point is to not force the driver to store the `user_data` pointer and then pass it to the callback (and force to user to provide some value for this pointer), when in (perhaps) most cases it won't be used. If there's a need for such identification of the object related to the callback, one could do it on the application side by using a dedicated variable. Even in the case when a common callback function is to be used for different counters (and checking `dev` pointer would be an ineffective way of identifying the callback cause), one can define separate callbacks that would call this common function with a proper parameter.

Reply:
Added additional note: "Not valid if callback is NULL." but i still think that it should stay. Sooner or later there is always a need to get user handle somehow in the callback. alarm_set do that through non-volatile requirement to the callback, here this is the best way imo.